src/Sarif.Sarifer/BackgroundAnalysisService.cs

```diff
@@ -17,37 +18,111 @@ namespace Microsoft.CodeAnalysis.Sarif.Sarifer
     [Export(typeof(IBackgroundAnalysisService))]
     internal class BackgroundAnalysisService : IBackgroundAnalysisService
     {
-#pragma warning disable CS0649 // Filled in by MEF
-#pragma warning disable IDE0044 // Assigned by MEF
+#pragma warning disable IDE0044, CS0649 // Provided by MEF
         [ImportMany]
         private IEnumerable<IBackgroundAnalyzer> analyzers;
-#pragma warning restore IDE0044
-#pragma warning restore CS0649
+
+        [ImportMany]
+        private IEnumerable<IBackgroundAnalysisSink> sinks;
+#pragma warning restore IDE0044, CS0649
 
         /// <inheritdoc/>
         public async Task AnalyzeAsync(string path, string text, CancellationToken cancellationToken)
         {
-            var tasks = new List<Task>(this.analyzers.Count());
+            var tasks = new List<Task<Stream>>(this.analyzers.Count());
 
             foreach (IBackgroundAnalyzer analyzer in this.analyzers)
             {
                 tasks.Add(analyzer.AnalyzeAsync(path, text, cancellationToken));
             }
 
-            await Task.WhenAll(tasks).ConfigureAwait(continueOnCapturedContext: false);
+            Stream[] streams = await Task.WhenAll(tasks).ConfigureAwait(continueOnCapturedContext: false);
+
+            try
+            {
+                await this.WriteStreamsToSinksAsync(path, streams, cleanAll: false).ConfigureAwait(continueOnCapturedContext: false);
+            }
+            finally
+            {
+                DisposeStreams(streams);
+            }
         }
 
         /// <inheritdoc/>
         public async Task AnalyzeAsync(string logId, IEnumerable<string> targetFiles, CancellationToken cancellationToken)
         {
-            var tasks = new List<Task>(this.analyzers.Count());
+            var tasks = new List<Task<Stream>>(this.analyzers.Count());
 
             foreach (IBackgroundAnalyzer analyzer in this.analyzers)
             {
-                tasks.Add(analyzer.AnalyzeAsync(logId, targetFiles, cancellationToken));
+                tasks.Add(analyzer.AnalyzeAsync(targetFiles, cancellationToken));
+            }
+
+            Stream[] streams = await Task.WhenAll(tasks).ConfigureAwait(continueOnCapturedContext: false);
+            try
+            {
+                await this.WriteStreamsToSinksAsync(logId, streams, cleanAll: true).ConfigureAwait(continueOnCapturedContext: false);
```

Comment:
>streams [](start = 59, length = 7)

Now some of them might be null, so perhaps

```
Stream streams = await Task.WhenAll(tasks).ConfigureAwait(continueOnCapturedContext: false).Where(s => s != Stream.Null).ToArray();
```

Reply:
in the method `WriteStreamsToSinksAsync` i'm checking for that.

---
In reply to: [535655681](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535655681) [](ancestors = 535655681)

Reply:
You also have to check in `DisposeStreams`.

So rather than checking in both places, you can filter out the `null` ones here (as in my code fragment above) and then you never have to worry about t again.

---
In reply to: [535657988](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535657988) [](ancestors = 535657988,535655681)

Reply:
the dispose of a Stream.null is a Noop: [https://referencesource.microsoft.com/#mscorlib/system/io/stream.cs,b7f982f75278c3c9](https://referencesource.microsoft.com/#mscorlib/system/io/stream.cs,b7f982f75278c3c9). I didn't do the where/toarray, because we will create a new array, so more objects used.

---
In reply to: [535661666](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535661666) [](ancestors = 535661666,535657988,535655681)

Reply:
Nice! TIL. I withdraw my objection.

---
In reply to: [535665533](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535665533) [](ancestors = 535665533,535661666,535657988,535655681)